pallets/sto/src/lib.rs

```diff
@@ -0,0 +1,333 @@
+}
+
+/// Single price tier of a `Fundraiser`.
+/// Similar to a `PriceTier` but with an extra field `remaining` for tracking the amount available in a tier.
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct FundraiserTier<Balance> {
+    inner: PriceTier<Balance>,
+    remaining: Balance,
+}
+
+impl<Balance: From<u8>> Into<FundraiserTier<Balance>> for PriceTier<Balance> {
+    fn into(self) -> FundraiserTier<Balance> {
+        FundraiserTier {
+            inner: self,
+            remaining: Balance::from(0),
+        }
+    }
+}
+
+pub trait Trait:
+    frame_system::Trait
+    + CommonTrait
+    + IdentityTrait
+    + SettlementTrait
+    + TimestampTrait
+    + PortfolioTrait
+{
+    /// The overarching event type.
+    type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+}
+
+decl_event!(
+    pub enum Event<T>
+    where
+        Balance = <T as CommonTrait>::Balance,
+        Moment = <T as TimestampTrait>::Moment,
+    {
+        /// A new fundraiser has been created
+        /// (primary issuance agent, fundraiser)
+        FundraiserCreated(IdentityId, Fundraiser<Balance, Moment>),
+        /// An investor invested in the fundraiser
+        /// (offering token, raise token, offering_token_amount, raise_token_amount, fundraiser_id)
+        FundsRaised(IdentityId, Ticker, Ticker, Balance, Balance, u64),
+    }
+);
+
+decl_error! {
+    /// Errors for the Settlement module.
+    pub enum Error for Module<T: Trait> {
+        /// Sender does not have required permissions
+        Unauthorized,
+        /// An arithmetic operation overflowed
+        Overflow,
+        /// Not enough tokens left for sale
+        InsufficientTokensRemaining,
+        /// Fundraiser not found
+        FundraiserNotFound,
+        /// Fundraiser is frozen
+        FundraiserFrozen,
+        // Interacting with a fundraiser past the end `Moment`.
+        FundraiserExpired,
+        // Interacting with a fundraiser before the start `Moment`.
+        FundraiserNotStated,
+        // Using an invalid venue
+        InvalidVenue,
+        // Using an invalid portfolio
+        InvalidPortfolio,
+    }
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as StoCapped {
+        /// All fundraisers that are currently running. (ticker, fundraiser_id) -> Fundraiser
+        Fundraisers get(fn fundraisers): double_map hasher(blake2_128_concat) Ticker, hasher(twox_64_concat) u64 => Fundraiser<T::Balance, T::Moment>;
+        /// Total fundraisers created for a token
+        FundraiserCount get(fn fundraiser_count): map hasher(twox_64_concat) Ticker => u64;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// Create a new offering. A fixed amount of pre-minted tokens are put up for sale at the specified tiered rate.
```

Comment:
If we can elaborate in the documentation re. all the errors and conditions it would be good I think.